src/lib/reasoners/rel_utils.ml

```diff
@@ -193,126 +193,179 @@ end = struct
 
-      It provides an interface to register some semantic values of interest, and
-      for applying a callback when the representative of those registered values
-      change.
-  *)
+  val subst : Explanation.t -> X.r -> X.r -> t -> t
+  (** [subst ex p v cs] replaces all the instaces of [p] with [v] in the
+      constraint.
+
+      Use this to ensure that the representation is always normalized.
+
+      The explanation [ex] justifies the equality [p = v]. *)
+
+  val fold_leaves : (X.r -> 'a -> 'a) -> t -> 'a -> 'a
+
+  type domain
+
+  val propagate : t -> domain -> domain
+  (** [propagate c dom] propagates the constraints [c] in [d] and returns the
+      new domains. *)
+
+end
 
+module Constraints_Make(Constraint : Constraint) : sig
   type t
-  (** The type of congruences. *)
+  (** The type of constraint sets. A constraint sets records a set of
+      constraints that applies to semantic values, and remembers which
+      constraints are associated with each semantic values.
+
+      It is used to only propagate constraints involving semantic values whose
+      associated domain has changed.
+
+      The constraint sets are expected to keep track of *class representatives*,
+      i.e.  normal forms wrt the `Uf` module, in which case we say the
+      constraint set is *normalized*. Use `subst` to ensure normalization. *)
+
+  val pp : t Fmt.t
+  (** Pretty-printer for constraint sets. *)
 
   val empty : t
-  (** The empty congruence. *)
+  (** Returns an empty constraint set. *)
 
-  val add : X.r -> t -> t
-  (** [add r t] registers the semantic value [r] in the congruence. *)
+  val subst : Explanation.t -> X.r -> X.r -> t -> t
+  (** [subst ex p v cs] replaces all the instances of [p] with [v] in the
+      constraints.
 
-  val remove : X.r -> t -> t
-  (** [remove r t] unregisters the semantic value [r] from the congruence.
+      Use this to ensure that the representation is always normalized.
 
-      Note that if substitutions have been applied to the congruence after a
-      value has been added, those same substitutions must be applied to the
-      semantic value prior to calling [remove], or [Invalid_argument] will be
-      raised.
+      The explanation [ex] justifies the equality [p = v]. *)
 
-      @raise [Invalid_argument] if [r] is not a registered semantic value. *)
+  val add : t -> Constraint.t -> t
+  (** [add c cs] adds the constraint [c] to [cs]. *)
 
-  val subst : X.r -> X.r -> t -> (X.r -> X.r -> 'a -> 'a) -> 'a -> t * 'a
-  (** [subst p v t f x] performs a local congruence closure of the
-      substitution [p -> v].
+  val fold_fresh : (Constraint.t -> 'a -> 'a) -> t -> 'a -> t * 'a
+  (** [fold_fresh f cs acc] folds [f] over the fresh constraints in [cs].
 
-      More precisely, it will fold [f] over the pairs [(rr, nrr)] such that:
-      - [rr] was registered in the congruence
-      - [nrr] is [X.subst p v rr]
+      Fresh constraints are constraints that were never propagated yet. *)
```

Comment:
We should mention that the fresh constraint queue is flushed by this iterator.

Reply:
I will rename this function as `propagate_fresh` and only expose `fold_fresh Constraint.propagate` and mention this in the doc of that function.